proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
+          apply (rule_tac Q="\<lambda>rv'. (case buf of None \<Rightarrow> \<lambda>_. True
+                                        | Some x \<Rightarrow> valid_ipc_buffer_ptr' x) and
+                                       invs' and sc_at' scp and cur_tcb'"
+                 in hoare_strengthen_post[rotated])
+           apply clarsimp
+          apply (wpsimp wp:hoare_case_option_wp)
+         apply clarsimp
+         apply (clarsimp simp: invs_def valid_state_def valid_pspace_def valid_sched_def)
+         apply (fastforce simp: sc_tcb_sc_at_def obj_at_def)
+        apply (clarsimp simp: invs'_def valid_pspace'_def obj_at'_def projectKOs)
+       apply wpsimp
+        apply (rule_tac Q="\<lambda>x s. (case buf of None \<Rightarrow> \<lambda>_. True
+                                | Some x \<Rightarrow> in_user_frame x)
+                                s \<and>
+                               invs s \<and>
+                               valid_list s \<and>
+                               valid_sched s \<and>
+                               bound_yt_tcb_at ((=) None) (cur_thread s) s \<and>
+                               sc_tcb_sc_at
+                                (\<lambda>sctcb. \<exists>t. sctcb = Some t \<and> t \<noteq> cur_thread s)
+                                scp s"
+               in hoare_strengthen_post[rotated])
+         apply (clarsimp simp: sc_yf_sc_at_def obj_at_def)
+        apply (wpsimp wp: hoare_case_option_wp complete_yield_to_invs)
+        apply ((wpsimp wp: complete_yield_to_sc_tcb_sc_at | wps)+)
+       apply (clarsimp split: if_split simp: sc_yf_sc_at_def obj_at_def)
+      apply (wpsimp wp: hoare_case_option_wp schedContextCompleteYieldTo_invs')
+     apply wpsimp+
+   apply (fastforce simp: sc_tcb_sc_at_def obj_at_def is_sc_obj
+                   elim!: valid_sched_context_size_objsI[OF invs_valid_objs])
+  apply clarsimp
+  done
+
+crunches sched_context_unbind_ntfn, sched_context_unbind_all_tcbs
+  for sc_at[wp]: "sc_at scp"
+  (wp: crunch_wps)
+
+crunches schedContextUnbindNtfn, schedContextUnbindAllTCBs
+  for sc_at'[wp]: "sc_at' scp"
+  (wp: crunch_wps)
+
+lemma invokeSchedContext_corres:
   "sc_inv_rel sc_inv sc_inv' \<Longrightarrow>
    corres (=)
-          (einvs and valid_sched_context_inv sc_inv and simple_sched_action and ct_active)
-          (invs' and sch_act_simple and valid_sc_inv' sc_inv' and ct_active')
+          (einvs and valid_sched_context_inv sc_inv and simple_sched_action
+           and current_time_bounded 2)
+          (invs' and sch_act_simple and valid_sc_inv' sc_inv')
           (invoke_sched_context sc_inv)
           (invokeSchedContext sc_inv')"
   apply (simp add: invoke_sched_context_def invokeSchedContext_def)
-  (* most of the next layer down should go into SchedContext_R, because some of these are
-     reused in Finalise and IpcCancel *)
-  sorry
+  apply (cases sc_inv; cases sc_inv'; clarsimp simp: cap_relation_def)
+      defer
```

Comment:
Do you still need this `defer`? Looking at the proof of the final subgoal, it doesn't seem to be doing anything special now that would require it to be left until the end.

Reply:
I should remove this. This was for moving `setConsumed` case to the last to deal with it later.